arroyo/backends/kafka/consumer.py

```diff
@@ -431,29 +471,33 @@ def __validate_offsets(self, offsets: Mapping[Partition, int]) -> None:
         if invalid_offsets:
             raise ConsumerError(f"invalid offsets: {invalid_offsets!r}")
 
+    def __assign(self, offsets: Mapping[Partition, int]) -> None:
+        self.__validate_offsets(offsets)
+        self.__consumer.assign(
+            [
+                ConfluentTopicPartition(partition.topic.name, partition.index, offset)
+                for partition, offset in offsets.items()
+            ]
+        )
+        self.__offsets.update(offsets)
+
+    def __incremental_assign(self, offsets: Mapping[Partition, int]) -> None:
+        self.__validate_offsets(offsets)
+        self.__consumer.incremental_assign(
+            [
+                ConfluentTopicPartition(partition.topic.name, partition.index, offset)
+                for partition, offset in offsets.items()
+            ]
+        )
+        self.__offsets.update(offsets)
+
     def __seek(self, offsets: Mapping[Partition, int]) -> None:
         self.__validate_offsets(offsets)
 
-        if self.__state is KafkaConsumerState.ASSIGNING:
-            # Calling ``seek`` on the Confluent consumer from an assignment
-            # callback will throw an "Erroneous state" error. Instead,
-            # partition offsets have to be initialized by calling ``assign``.
-            self.__consumer.assign(
-                [
-                    ConfluentTopicPartition(
-                        partition.topic.name, partition.index, offset
-                    )
-                    for partition, offset in offsets.items()
-                ]
+        for partition, offset in offsets.items():
+            self.__consumer.seek(
+                ConfluentTopicPartition(partition.topic.name, partition.index, offset)
```

Comment:
Why this change ?

Reply:
I found it pretty unintuitive that the same `__seek` implementation was being called from both `seek()` and the assign callback then checking the KafkaConsumerState to effectively determine which one was making the call. Now assigning function is separate and never calls this `__seek` function anymore.

Reply:
Also, now this method is only called by the `seek` public method. Why not inlining it there ?

Reply:
I was avoiding touching a lot of code earlier, but yes it is better inline since no one else calls it now. Updated.